File: LeetCode/704_BinarySearch.py
# https://leetcode.com/problems/binary-search/

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # # v1
        # lb=0 #lower boundary
        # hb=len(nums)-1
        # if target<nums[0] or target>nums[-1]:
        #     return -1
        # for i in range(len(nums)):
        #     half=(hb+lb)//2
        #     if nums[half]==target:
        #         return half
        #     elif nums[half]<target:
        #         lb=half
        #         if nums[half+1]>target:
        #             return -1
        #     elif nums[half]>target:
        #         hb=half
        #         if nums[half-1]<target:
        #             return -1

        # v2
        lb,hb=0,len(nums)-1
        while lb<=hb:
            pivot=(lb+hb)//2
            if nums[pivot]==target:
                return pivot
            if target<nums[pivot]:
                hb=pivot-1
            else:
                lb=pivot+1
        return -1

File: LeetCode/704_BinarySearch.go
// https://leetcode.com/problems/binary-search/
// 2022年07月02日 10:24:07

import "fmt"
func search(nums []int, target int) int {
    l,h:=0,len(nums)-1
    for l<=h {
        mid:=(l+h)/2
        if nums[mid]<target {
            l=mid+1
        }else if nums[mid]>target{
            h=mid-1
        }else{
            return mid
        }
    } 
    return -1
}